Added RedisStorage class to support storing activity on Redis

pablete opened the pull request:
I have found the current sql stores (msyql / postgress) are very slow when the number of trials increases to more than 300... or so.

I am interested in a team or standalone UI for distributed use. Typically, queries for visualization take between 3/5 minutes to render a graph which is not suitable for web apps or dashboards.

While working on this standalone UI, I decided to contribute with the RedisStorage class so others can benefit of having a faster (but also simpler) storage.

pablete commented:
Forgot to install redis-server on circleci. Working on it

hvy commented:
Thanks a lot for this welcomed PR. I'll definitely have a look at it!

hvy commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+
+        self.redis.set("trial_id:{:010d}:frozentrial".format(trial_id), pickle.dumps(trial))
+
+    def del_trial(self, trial_id):
```
How about prefixing this method with an underscore making it obviously private?

pablete replied:
Right. Prefixed all private methods with _ make the intention clearer. DONE

hvy commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+    from typing import Optional  # NOQA
+
+
+class RedisStorage(base.BaseStorage):
```
To allow multiple processes accessing the same Redis database (similar to our other database. See https://optuna.readthedocs.io/en/latest/tutorial/distributed.html) we might have to rely on Redis' `WATCH` (e.g. https://redis-py.readthedocs.io/en/latest/index.html#redis.Redis.pipeline, https://redis-py.readthedocs.io/en/latest/index.html#redis.Redis.transaction). We could tackle that in a separate PR since this change is already quite big. In that case, we might want to mark this feature as experimental.

pablete replied:
Added pipeline blocks in places where multiple updates where performed. That will pipeline all writes as one atomic transaction.

hvy commented:
By the way, I ran some quick benchmarks
https://gist.github.com/hvy/8c08604ba68f1e4b115b09e18c6c39a7.

- `get_all_study_summaries` is indeed significantly faster (assuming you don't rely on any cache)
- As for optimization, it seems like the running time per trial increases quite rapidly as you run more trials. Is this consistent with what you experience?

hvy commented:
A follow up to this PR includes fixing the optuna cli to support this storage e.g. allowing deletions of studies via `$ optuna delete-study --study-name "foobar" --storage "redis://localhost:6379/0"` (see https://github.com/optuna/optuna/blob/master/optuna/cli.py#L94-L97).

hvy commented:
`head`-ed cProfile output optimizing 10 trials, after around 600 trials have already been finished. 

```
         9645112 function calls (8777491 primitive calls) in 6.637 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2373/1    0.019    0.000    6.639    6.639 {built-in method builtins.exec}
        1    0.000    0.000    6.639    6.639 quadratic_simple.py:1(<module>)
        1    0.000    0.000    4.870    4.870 study.py:254(optimize)
        1    0.000    0.000    4.870    4.870 study.py:569(_optimize_sequential)
       10    0.000    0.000    4.870    0.487 study.py:618(_run_trial_and_callbacks)
       10    0.000    0.000    4.870    0.487 study.py:633(_run_trial)
       40    0.024    0.001    4.114    0.103 redis_storage.py:481(get_all_trials)
      388    0.003    0.000    3.084    0.008 __init__.py:1(<module>)
       30    0.003    0.000    2.800    0.093 study.py:120(get_trials)
    25930    0.035    0.000    2.659    0.000 redis_storage.py:458(get_trial)
    26392    0.067    0.000    2.352    0.000 client.py:871(execute_command)
    26055    0.014    0.000    2.151    0.000 client.py:1575(get)
...
```

It seems like `get_all_trials` (which is called 4 times here per trial) is dominating the trial cumtime time.

pablete commented on optuna/storages/__init__.py:
```diff
@@ -11,6 +12,9 @@ def get_storage(storage):
         return InMemoryStorage()
     if isinstance(storage, str):
-        return RDBStorage(storage)
+        if storage[:5] == 'redis':
```
DONE

pablete commented on setup.py:
```diff
@@ -39,6 +39,7 @@ def get_install_requires():
         'numpy',
         'scipy!=1.4.0',
         'sqlalchemy>=1.1.0',
+        'redis',
```
DONE

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+
+    """
+
+    def __init__(self, redis_url='redis://localhost:6379/0', testing_environment=False):
```
DONE

pablete commented on docs/source/reference/storages.rst:
```diff
@@ -4,3 +4,4 @@ Storages
 ========
 
 .. autoclass:: RDBStorage
+.. autoclass:: optuna.storages.redis_storage.RedisStorage
```
I am afraid it does not work by simply using the RedisStorage class name...

hvy replied:
Hm, what kind of failure are you experiencing?

pablete replied:
Changed it. it works now.

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+    """
+
+    def __init__(self, redis_url='redis://localhost:6379/0', testing_environment=False):
+        # type: (str, bool) -> None
```
DONE

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+
+    """
+
+    def __init__(self, redis_url='redis://localhost:6379/0', testing_environment=False):
```
Ok. Refactored the code to reflect the absence of this flag.

hvy replied:
Thanks, this makes `__init__` simpler. However, the new method is still public. I wonder if we shouldn't just create a function in the unit test instead that does precisely what `init_with_flush` does, (but with the specific DB URL).

pablete replied:
Refactored init_with_flush out for RedisStorage into test_storage.py.

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+    def create_new_study(self, study_name=None):
+        # type: (Optional[str]) -> int
+
+        study_name_key = "study_name:{}:study_id".format(study_name)
```
I prefer the keys to match the snake case (underscore) present in python.  The : colon to separate keys:id:subkey is what I respected.

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+            self.redis.set("trial_counter", -1)
+        trial_id = self.redis.incr("trial_counter", 1)
+        trial.system_attrs['_number'] = trial_id
+        trial.number = trial_id
```
Correct. Good point. I overlooked this since I was based on the InMemoryStorage class.
DONE

pablete commented on optuna/storages/redis_storage.py:
```diff
@@ -0,0 +1,511 @@
+
+    """
+
+    def __init__(self, redis_url='redis://localhost:6379/0', testing_environment=False):
```
Changed to url. DONE

pablete commented on setup.cfg:
```diff
@@ -12,3 +12,11 @@ exclude = venv,build
+ignore_missing_imports = True
+
+[mypy-optuna.storages.redis_storage]
+strict_optional = False
```
strict_optional = False made the pull request pass the tests, otherwise it complains about 

Argument 1 to "loads" has incompatible type "Optional[bytes]"; expected "bytes"

in all of redis get statements.
Perhaps you can cast some light on that error which I do not have much experience with.

hvy replied:
Sorry for my earlier brief comment. I think the errors derive from `pickle.loads` and `float`? mypy complains when you call  `pickle.loads` with arguments that may be `None` (type hinted to be `Optional`).  `pickle.loads` apparently does not expect that. Similarly for `float`. What you can do is to create temporaries and use `assert`s or explicitly check the type using e.g. `if`s, to let mypy know that these values cannot be `None`. 

**Current**
```python
pickle.loads(self.redis.get(foo))  # `self.redis.get` might return `None`?
```
**Possible solution**
```python
bar = self.redis.get(foo)
assert bar is not None
# or... 
# if bar is None:
#     ....  # e.g. raise an error here or call `pickle.loads` in the `else` clause.
pickle.loads(bar)  
```

hvy replied:
You can run `mypy` locally, e.g. 

```bash
mypy optuna/storages/redis_storage.py
```

pablete replied:
Thanks..addressed all Optional[bytes] coming from redis

pablete commented:
Thanks @hvy for the extensive review. 
I have tried to comply with all of your requests. 
Thanks to spend the time on building a benchmark comparison. It will serve me as a template to improve the performance. But as a general rule, If you want the reads to be fast you have to sacrifice some of the time in the writes. That is why the time per trial insert grows as a function of the number of trials. I'll work on tuning this to get better performance in a separate PR.

pablete commented:
Will work on this
`$ optuna delete-study --study-name "foobar" --storage "redis://localhost:6379/0"`

shortly...

hvy commented:
This is another detail but we might want to rename `redis_storage.py` to `redis.py` just to be consistent with `base.py` (`BaseStorage`) and `in_memory.py` (`InMemoryStorage`).

pablete commented on .circleci/config.yml:
```diff
@@ -72,7 +72,7 @@ jobs:
           command: |
             . venv/bin/activate
-            mypy --disallow-untyped-defs --ignore-missing-imports .
+            mypy .
```
Do you prefer that I reverse this change?

hvy replied:
If you don't mind creating a separate PR for that, especially as this PR is already quite big. I'd assume we can merge that one rather quickly.

pablete replied:
Reverted. Now this change lives in #985

pablete commented:
Worked on the cli, that now supports operations against redis://localhost:port/db style urls.

One major pending is to improve the insert (new trial) performance, If I can squeeze some time tomorrow I will include it in this PR.

hvy commented on docs/source/reference/storages.rst:
```diff
@@ -4,3 +4,5 @@ Storages
 
 .. autoclass:: RDBStorage
+.. autoclass:: RedisStorage
+
```
A nit.
```suggestion
```

hvy commented:
@pablete thanks for addressing the comments so quickly including the cli and pipelines. I'll try to take a second look tomorrow or during the weekend. 

I just want to let you know that we have a PR working on moving the `number` attribute out from the `system_attrs` but into the database schema (https://github.com/optuna/optuna/pull/939/). The way it is now is mostly due to historical reasons AFAIK. This might affect this PR e.g. the insertion of new trials that you mention above. I'm not sure what the best way to proceed though is, as it could depend on the order of these PRs. @c-bata do you have any comments on this?

hvy commented:
Leaving some thought being very open for discussions. I was considering alternatives to pickle e.g. JSON (I'm not sure this possible). pickle might be suboptimal in terms of performance (speed), though this might be somewhat mitigated by `cPickle`. Another perspective is the tight coupling to the Python programming language. @c-bata might have some comments on this as well as the author of [Goptuna](https://github.com/c-bata/goptuna).

pablete commented:
On the subject of **pickle vs json**, I think pickle gave me a straightforward way to serialize objects in optuna.structs without modifying them.  If in the future, these primitive optuna.structs get their one serialize/unserialize mechanisms, me or anyone  can easily change them in redis.py

pablete commented:
In reference of #939 (moving the number attribute out from the system_attrs but into the database schema) It does not affect redis.py I guess. I am already using optuna.structs.FrozenTrial that includes number as a part of the trial itself. I just serialize the whole primitive struct and toss it into redis database.

c-bata commented:
> I just want to let you know that we have a PR working on moving the number attribute out from the system_attrs but into the database schema (#939). The way it is now is mostly due to historical reasons AFAIK. This might affect this PR e.g. the insertion of new trials that you mention above. I'm not sure what the best way to proceed though is, as it could depend on the order of these PRs. @c-bata do you have any comments on this?

There are no serious conflicts between #939 and this pull request. I guess some test cases will be failed if the following line is removed before merging #939. I think it should be better to put a todo comment.

https://github.com/optuna/optuna/pull/974/files#diff-fd41d65a6ae629878ca001ce9486c0ebR298

---

> On the subject of pickle vs json, I think pickle gave me a straightforward way to serialize objects in optuna.structs without modifying them. If in the future, these primitive optuna.structs get their one serialize/unserialize mechanisms, me or anyone can easily change them in redis.py

I think your opinion is reasonable :+1:

c-bata commented on tests/storages_tests/test_storages.py:
```diff
@@ -93,9 +94,20 @@
-    'storage_init_func', [InMemoryStorage, lambda: RDBStorage('sqlite:///:memory:')])
+    'storage_init_func', [InMemoryStorage,
+                          lambda: RDBStorage('sqlite:///:memory:'),
+                          lambda: redis_with_flushdb()])
```
I think you don't need to use a lambda function.

```suggestion
                          redis_with_flushdb])
```

pablete replied:
You are right. Thanks

hvy commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,577 @@
+        trial_id = self.redis.incr("trial_counter", 1)
+        trial_number = self.redis.incr("study_id:{:010d}:trial_number".format(study_id))
+        # TODO(pablete): Remove number from system_attrs after adding TrialModel.number.
+        trial.system_attrs['_number'] = trial_number
```
Sorry could you remove this line? #939 was just merged 🙂

pablete replied:
Removed.

pablete commented:
> May I have the reason why you use `threading.RLock` in the RedisStorage? I quickly checked the code but I couldn't understand what objects you want to protect.
> 
> At least, redis-py's client objects are thread-safe.
> 
> > Redis client instances can safely be shared between threads.
> > https://github.com/andymccurdy/redis-py

You are absolutely right, I have left the reentrant locks there becase I based the redis storage on the in_memory storage. Turns out that they are not needed.

hvy commented:
Could you try to fix the commit history and the conflict? Sorry having it long-running I'll try to review this PR as soon as possible.

pablete commented:
Rebased. This branch now is up to date with master.
Thanks!

hvy commented:
Thanks @pablete for the rebase! It seems however that the commit history is still a bit messy (this PR contains 195 commits). Could you possibly clean the diff as well?

pablete commented:
Hi @hvy. I am not sure what you meant by "commit history is a bit messy" what does it mean, you want me to squash the commits into one? I do not know if I can revert the "rebase" I performed.
I am only using the web ui to address this, how can I clean "the diff" ?

Thanks

hvy commented:
Seems like you got it solved. Thanks. 

By the way, we recently adopted `black` as the formatter so some failures might occur in  the CI. As you've solved the conflicts already, you probably just have to run black locally. 

```bash
black <path to Python file/directory> --line-length 99
```
C.f. https://github.com/optuna/optuna/pull/1030#issue-388996877

pablete commented:
Done, compact, and checked with black.

hvy commented on setup.cfg:
```diff
@@ -9,8 +9,3 @@ exclude = venv,build
-# This section is for mypy.
-[mypy]
-disallow_untyped_defs = True
-ignore_missing_imports = True
```
Ditto. 
> You've maybe by mistake reverted #985 ?

pablete replied:
Reverting.

hvy commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,570 @@
+    from typing import Optional  # NOQA
+
+
+class RedisStorage(base.BaseStorage):
```
Could you wrap this class using the `experimental` decorator? I'd like to merge this PR rather soon although other maintainers/committers haven't yet had the time to take a proper look at the changes in my understanding. I basically think the internals LGTM.

pablete replied:
Added @experimental("1.2.0") as suggested.

pablete commented on .circleci/config.yml:
```diff
@@ -72,7 +72,7 @@ jobs:
           command: |
             . venv/bin/activate
-            mypy .
+            mypy --disallow-untyped-defs --ignore-missing-imports .
```
You are right, my bad. Reverting.

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,559 @@
+
+        self.redis = redis.Redis.from_url(url)
+
+    def __getstate__(self):
```
Removed

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,570 @@
+import copy
+from datetime import datetime
+import pickle
+import redis
```
Done

pablete commented on tests/storages_tests/test_storages.py:
```diff
@@ -71,9 +72,19 @@
+def redis_with_flushdb():
+    # type: () -> BaseStorage
+
+    storage = RedisStorage("redis://localhost:6379/15")
```
Included fakeredis. Good point.

pablete commented:
Addressed all comments. Hope this gets merged soon, so I can work on additional PRs ;)

sile commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,562 @@
+    from typing import Optional  # NOQA
+
+
+@experimental("1.2.0")
```
Please let me suggest the following change because the next release version will be `1.3.0`.
```suggestion
@experimental("1.3.0")
```

hvy commented on setup.py:
```diff
@@ -35,6 +35,7 @@ def get_install_requires() -> List[str]:
         "colorlog",
         "joblib",
         "numpy",
+        "redis",
```
I actually think that `redis` is a rather trivial dependency compared to e.g. `alembic` which is used for our RDB storage. `redis` only uses the standard library.

sile replied:
I agree with @c-bata because `RedisStorage` is still an experimental feature. I think that we can make `redis` as a required dependency when the feature become stable.

pablete replied:
So, you want me to make it extra requirement for now, and move it as requires dependency later?

pablete replied:
Added an experimental section in setup.py

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,562 @@
+
+import redis
+
+from optuna import distributions, exceptions  # NOQA
```
Done.

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,562 @@
+from datetime import datetime
+import pickle
+
+import redis
```
Fair enough, made it similar with _is_available flag.

pablete commented:
Addressed all comments. RedisStorage decorate experimental(1.3.0). Make redis optional via experimental section in setup.py and _is_available flag on import, as suggested.

sile commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+        self.redis.set("study_id:{:010d}:trial_number".format(study_id), -1)
+
+        if study_name is None:
+            study_name = "{}{:010d}".format(DEFAULT_STUDY_NAME_PREFIX, study_id)
```
I added the following unit test to `tests/storages_tests/test_storages.py` on my local environment.
```python
@parametrize_storage
def test_create_and_delete_anonymous_study(storage_init_func):
    # type: (Callable[[], BaseStorage]) -> None

    storage = storage_init_func()
    study_id = storage.create_new_study()
    study_name = storage.get_study_name_from_id(study_id)
    assert study_id == storage.get_study_id_from_name(study_name)
```

This test was failed as follows:
```python
$ pytest tests/storages_tests/test_storages.py
...
=================================================================================================================== FAILURES ===================================================================================================================__________________________________________________________________________________________ test_create_and_delete_anonymous_study[redis_with_flushdb] __________________________________________________________________________________________

storage_init_func = <function redis_with_flushdb at 0x7fe733fbb620>

    @parametrize_storage
    def test_create_and_delete_anonymous_study(storage_init_func):
        # type: (Callable[[], BaseStorage]) -> None

        storage = storage_init_func()
        study_id = storage.create_new_study()
        study_name = storage.get_study_name_from_id(study_id)
>       assert study_id == storage.get_study_id_from_name(study_name)

tests/storages_tests/test_storages.py:112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <optuna.storages.redis.RedisStorage object at 0x7fe711f7b748>, study_name = 'no-name-0000000000'

    def get_study_id_from_name(self, study_name):
        # type: (str) -> int

        if not self.redis.exists("study_name:{}:study_id".format(study_name)):
>           raise ValueError("No such study {}.".format(study_name))
E           ValueError: No such study no-name-0000000000.

optuna/storages/redis.py:218: ValueError
```

I think that the reassignment of `key_study_name` (i.e., `key_study_name = "study_name:{}:study_id".format(study_name)`) is needed after the updating of `study_name`.

pablete replied:
Good catch!. Sorted it out. Added the test as well.

sile replied:
👍

pablete commented on tests/storages_tests/test_storages.py:
```diff
@@ -71,9 +73,20 @@
+def redis_with_flushdb():
+    # type: () -> BaseStorage
+
+    # the redis url is not going to be used, but it has to be valid
```
DONE

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+            >>> study.optimize(objective)
+
+    Args:
+        url: URL of the redis storage. password and db are optional
```
DONE

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+                "(The actual import error is as follows: " + str(_import_error) + ")"
+            )
+
+        self.redis = redis.Redis.from_url(url)
```
Addressed. Renamed to _redis

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+            raise exceptions.DuplicatedStudyError
+
+        if not self.redis.exists("study_counter"):
+            # we need the counter to start with 0
```
DONE

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+            # we need the counter to start with 0
+            self.redis.set("study_counter", -1)
+        study_id = self.redis.incr("study_counter", 1)
+        # we need the trial_number counter to start with 0
```
DONE

pablete commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,582 @@
+    Example:
+
+        We create an :class:`~optuna.storages.redis.RedisStorage` instance using
+        the given redis database URL
```
DONE

sile commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,586 @@
+        trial = self.get_trial(trial_id)
+        self.check_trial_is_updatable(trial_id, trial.state)
+
+        if state == structs.TrialState.RUNNING and trial.state != structs.TrialState.WAITING:
```
I think that this condition could be violated if, for example, the following scenario occurred.

- Assumption:
  - We are executing a distributed optimization with two worker processes, and
  - there is a trial with ID `10` and state `TrialState.WAITING`.
- (1) "A" calls `set_trial_state(10, TrialState.RUNNING)`, and is suspended just before line 354.
- (2) "B" calls `set_trial_state(10, TrialState.RUNNING)`, and is also suspended just before line 354.
  - Note that at the this time, both "A" and "B" have the exactly same `trial` .
- (3) "A" is resumed. Since the condition (line 354) is evaluated to `False`, this process sets the trial state to `RUNNING`.
- (4) "A" continues the evaluation of the trial, and eventually sets the state to `COMPLETE`.
- (5) After that, "B" is resumed. Since this process has the same `trial` as "A" in the step `(3)`, the condition is evaluated to `False`.
   - Then "B" changes the trial state to `RUNNING` from `COMPLETE` (not `WAITING`).
   - => This is unexpected behavior.

Obviously this is an artificial scenario, but I think that this and other methods that update DB state have potential risks like this.

sile replied:
To fix the above issue, it seems we need to rewrite the update methods to use `WATCH` redis command (i.e., optimistic locking mechanism) and retry if the watching keys were updated by other processes during a transaction. 
FYI, [The redis official doc about transaction](https://redis.io/topics/transactions) describes a typical flow to enable optimistic locking in Redis (the comments were added by me):
```
# EXAMPLE: An own implementation of INCR command (quoted from the redis official doc).
WATCH mykey     # (1) Watch target key
val = GET mykey # (2) Get the content of the key
val = val + 1   # (3) Update the content locally
MULTI           # (4) Start transaction
SET mykey $val  # (5) Enqueue the update request in the local command queue
EXEC            # (6) Execute the transaction atomically (if it failed, we need to retry)
```

Although we need to use the optimistic locking to fully support Optuna's distributed optimization, it could be a task requiring a bit large code modification. As the current implementation works fine when there is only single worker, I think that it's okay to merge this PR without distributed optimization support and address this issue in another PR.
What do you think of it?

pablete replied:
for 1) and 2) I do not understand why 2 processes will end up having the same trial ID = 10. 
line #308 of redis.py shows that the trial id is given by an atomic incr operation. This guarantees that calling create_new_trial will give you a different ID every time. (sequential integers). Perhaps there is a scenario that I do not currently see, where you reuse trials perhaps...

Also checking on rdb storage.py https://github.com/optuna/optuna/blob/master/optuna/storages/rdb/storage.py#L495-L497
has the same logic.

Perhaps is this is an issue we can merge this PR first. 
Then submit another PR tackling this concurrency scenario you describe having an integration test, but for all stores at once.

sile replied:
> for 1) and 2) I do not understand why 2 processes will end up having the same trial ID = 10.

Sorry for the lack of explanation. The scenario could occur when a user uses [Study.enqueue_trial](https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.Study.enqueue_trial) feature. To pop a queued trial (i.e., a trial with `WAITING` state) at the begging of a new trial round, multiple workers could call `set_trial_state` with the same trial ID simultaneously (see [here](https://github.com/optuna/optuna/blob/034cfdf1ff9cddcd9c0e5301f9a476181c5f2b41/optuna/study.py#L639) for the code).
Besides, though the detail has not been decided, we have a plan to expose `set_trial_state` to users (https://github.com/optuna/optuna/issues/685#issuecomment-554159519) via `Study` object. If it becomes a public API, users might call the method with the same ID from multiple processes.

> Perhaps is this is an issue we can merge this PR first.

Good. I'll take a final look this week and then approve this PR.

pablete replied:
Good to know. Thanks for the clarification.

sile commented on optuna/storages/redis.py:
```diff
@@ -0,0 +1,586 @@
+    from typing import Optional  # NOQA
+
+
+@experimental("1.3.0")
```
Since we released [v1.3.0](https://github.com/optuna/optuna/releases/tag/v1.3.0) today, please bump up the version:
```suggestion
@experimental("1.4.0")
```

hvy commented:
@pablete Would you like to implement the optimistic locking as well or should someone else take over?

pablete commented:
I can give it a shot. Trying to figure out a test case first, then I can jump on it.
I do not know if you work with issues, or project plan, or how you coordinate. (gitter?)

My main motivation is to contribute a complete dashboard. And perhaps a parallel version of TPE sampler.

hvy commented:
It'd be great if you could continue with the optimistic locking. 🙇 

And as for coordination, yes basically all of them. Are you concerned about the follow-up to this PR or the latter features mentioned? Generally, I prefer creating an issue here on GitHub to address follow-ups, discuss design or propose features. I'm feeling I might be missing your point, though. Gitter is an option but the downside is that it's not very convenient for tracking history.

hvy commented:
@pablete do you have any updates on the locking (the transaction described at the bottom here https://github.com/andymccurdy/redis-py#pipelines)? I'd like to continue with this and am wondering if we should raise a separate issue or find someone else to work on it if you're busy.
